File: backend/app/services/friend_service.py
from sqlalchemy.orm import Session
from sqlalchemy import func
from fastapi import HTTPException, status
from app.models.user import User
from app.models.friend_request import FriendRequest, FriendRequestStatus
from app.models.friend import Friend
from app.models.group import Group, GroupType
from app.models.group_member import GroupMember
from typing import List
import uuid


def create_friend_request(db: Session, sender: User, receiver_id: uuid.UUID) -> FriendRequest:
    """Create a friend request"""
    if sender.id == receiver_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot send friend request to yourself"
        )
    
    receiver = db.query(User).filter(User.id == receiver_id).first()
    if not receiver:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User not found"
        )
    
    existing_request = db.query(FriendRequest).filter(
        ((FriendRequest.sender_id == sender.id) & (FriendRequest.receiver_id == receiver_id)) |
        ((FriendRequest.sender_id == receiver_id) & (FriendRequest.receiver_id == sender.id))
    ).filter(FriendRequest.status == FriendRequestStatus.PENDING).first()
    
    if existing_request:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Friend request already exists"
        )
    
    existing_friendship = db.query(Friend).filter(
        ((Friend.user_id == sender.id) & (Friend.friend_id == receiver_id)) |
        ((Friend.user_id == receiver_id) & (Friend.friend_id == sender.id))
    ).first()
    
    if existing_friendship:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Already friends"
        )
    
    friend_request = FriendRequest(
        sender_id=sender.id,
        receiver_id=receiver_id,
        status=FriendRequestStatus.PENDING
    )
    
    db.add(friend_request)
    db.commit()
    db.refresh(friend_request)
    return friend_request


def accept_friend_request(db: Session, current_user: User, request_id: uuid.UUID) -> dict:
    """Accept a friend request and create DIRECT group"""
    friend_request = db.query(FriendRequest).filter(FriendRequest.id == request_id).first()
    
    if not friend_request:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Friend request not found"
        )
    
    if friend_request.receiver_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to accept this request"
        )
    
    if friend_request.status != FriendRequestStatus.PENDING:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Friend request already processed"
        )
    
    friend_request.status = FriendRequestStatus.ACCEPTED
    
    friend1 = Friend(
        user_id=friend_request.sender_id,
        friend_id=friend_request.receiver_id
    )
    friend2 = Friend(
        user_id=friend_request.receiver_id,
        friend_id=friend_request.sender_id
    )
    
    db.add(friend1)
    db.add(friend2)
    
    sender = db.query(User).filter(User.id == friend_request.sender_id).first()
    receiver = db.query(User).filter(User.id == friend_request.receiver_id).first()
    
    direct_group = Group(
        name=f"{sender.name} & {receiver.name}",
        created_by=current_user.id,
        type=GroupType.DIRECT,
        is_archived=False
    )
    
    db.add(direct_group)
    db.flush()
    
    member1 = GroupMember(group_id=direct_group.id, user_id=sender.id)
    member2 = GroupMember(group_id=direct_group.id, user_id=receiver.id)
    
    db.add(member1)
    db.add(member2)
    
    db.commit()
    db.refresh(friend_request)
    
    return {
        "message": "Friend request accepted",
        "direct_group_id": str(direct_group.id)
    }


def reject_friend_request(db: Session, current_user: User, request_id: uuid.UUID) -> FriendRequest:
    """Reject a friend request"""
    friend_request = db.query(FriendRequest).filter(FriendRequest.id == request_id).first()
    
    if not friend_request:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Friend request not found"
        )
    
    if friend_request.receiver_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to reject this request"
        )
    
    if friend_request.status != FriendRequestStatus.PENDING:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Friend request already processed"
        )
    
    friend_request.status = FriendRequestStatus.REJECTED
    db.commit()
    db.refresh(friend_request)
    return friend_request


def get_friends(db: Session, current_user: User) -> List[dict]:
    """Get list of friends for current user"""
    friends = db.query(Friend, User).join(
        User, Friend.friend_id == User.id
    ).filter(Friend.user_id == current_user.id).all()
    
    result = []
    for friend, user in friends:
        result.append({
            "user_id": friend.user_id,
            "friend_id": friend.friend_id,
            "friend_name": user.name,
            "friend_email": user.email,
            "created_at": friend.created_at
        })
    
    return result


def remove_friend(db: Session, current_user: User, friend_id: uuid.UUID) -> dict:
    """Remove a friend and delete the DIRECT group"""
    # Find the friendship (bidirectional)
    friendship1 = db.query(Friend).filter(
        Friend.user_id == current_user.id,
        Friend.friend_id == friend_id
    ).first()
    
    friendship2 = db.query(Friend).filter(
        Friend.user_id == friend_id,
        Friend.friend_id == current_user.id
    ).first()
    
    if not friendship1 and not friendship2:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Friendship not found"
        )
    
    # Find and delete the DIRECT group
    from app.models.group_member import GroupMember
    from app.models.expense import Expense
    from app.models.expense_item import ExpenseItem
    from app.models.split import Split
    
    # Find DIRECT group with both users
    direct_group = db.query(Group).filter(
        Group.type == GroupType.DIRECT
    ).join(
        GroupMember, Group.id == GroupMember.group_id
    ).filter(
        GroupMember.user_id.in_([current_user.id, friend_id])
    ).group_by(Group.id).having(
        func.count(GroupMember.user_id) == 2
    ).first()
    
    if direct_group:
        # Delete all related data
        expenses = db.query(Expense).filter(Expense.group_id == direct_group.id).all()
        for expense in expenses:
            items = db.query(ExpenseItem).filter(ExpenseItem.expense_id == expense.id).all()
            for item in items:
                db.query(Split).filter(Split.expense_item_id == item.id).delete()
            db.query(ExpenseItem).filter(ExpenseItem.expense_id == expense.id).delete()
        
        db.query(Expense).filter(Expense.group_id == direct_group.id).delete()
        db.query(GroupMember).filter(GroupMember.group_id == direct_group.id).delete()
        db.delete(direct_group)
    
    # Delete friendships
    if friendship1:
        db.delete(friendship1)
    if friendship2:
        db.delete(friendship2)
    
    db.commit()
    
    return {"message": "Friend removed successfully"}


def get_direct_group(db: Session, current_user: User, friend_id: uuid.UUID) -> dict:
    """Get the DIRECT group for a specific friend"""
    # Verify friendship exists
    friendship = db.query(Friend).filter(
        Friend.user_id == current_user.id,
        Friend.friend_id == friend_id
    ).first()
    
    if not friendship:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Friendship not found"
        )
    
    # Find DIRECT group with both users
    direct_group = db.query(Group).filter(
        Group.type == GroupType.DIRECT
    ).join(
        GroupMember, Group.id == GroupMember.group_id
    ).filter(
        GroupMember.user_id.in_([current_user.id, friend_id])
    ).group_by(Group.id).having(
        func.count(GroupMember.user_id) == 2
    ).first()
    
    if not direct_group:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="DIRECT group not found for this friendship"
        )
    
    return {
        "group_id": direct_group.id,
        "group_name": direct_group.name,
        "created_at": direct_group.created_at
    }

File: backend/src/main/java/com/snapsplit/service/FriendService.java
package com.snapsplit.service;

import com.snapsplit.dto.request.FriendRequestCreateRequest;
import com.snapsplit.dto.response.FriendRequestResponse;
import com.snapsplit.dto.response.FriendResponse;
import com.snapsplit.entity.*;
import com.snapsplit.enums.FriendRequestStatus;
import com.snapsplit.enums.GroupType;
import com.snapsplit.exception.BadRequestException;
import com.snapsplit.exception.ResourceNotFoundException;
import com.snapsplit.exception.UnauthorizedException;
import com.snapsplit.repository.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

/**
 * Friend service — equivalent to Python's friend_service.py
 */
@Service
public class FriendService {

    private final UserRepository userRepository;
    private final FriendRequestRepository friendRequestRepository;
    private final FriendRepository friendRepository;
    private final GroupRepository groupRepository;
    private final GroupMemberRepository groupMemberRepository;
    private final ExpenseRepository expenseRepository;
    private final ExpenseItemRepository expenseItemRepository;
    private final SplitRepository splitRepository;

    public FriendService(UserRepository userRepository,
            FriendRequestRepository friendRequestRepository,
            FriendRepository friendRepository,
            GroupRepository groupRepository,
            GroupMemberRepository groupMemberRepository,
            ExpenseRepository expenseRepository,
            ExpenseItemRepository expenseItemRepository,
            SplitRepository splitRepository) {
        this.userRepository = userRepository;
        this.friendRequestRepository = friendRequestRepository;
        this.friendRepository = friendRepository;
        this.groupRepository = groupRepository;
        this.groupMemberRepository = groupMemberRepository;
        this.expenseRepository = expenseRepository;
        this.expenseItemRepository = expenseItemRepository;
        this.splitRepository = splitRepository;
    }

    /**
     * Send a friend request by email.
     */
    @Transactional
    public FriendRequestResponse sendFriendRequest(UserEntity sender, FriendRequestCreateRequest request) {
        UserEntity receiver = userRepository.findByEmail(request.getFriendEmail())
                .orElseThrow(() -> new ResourceNotFoundException(
                        "User with email " + request.getFriendEmail() + " not found"));

        if (sender.getId().equals(receiver.getId())) {
            throw new BadRequestException("Cannot send friend request to yourself");
        }

        // Check existing pending request (bidirectional)
        boolean existsPending = friendRequestRepository.existsBySenderIdAndReceiverIdAndStatus(
                sender.getId(), receiver.getId(), FriendRequestStatus.PENDING) ||
                friendRequestRepository.existsBySenderIdAndReceiverIdAndStatus(
                        receiver.getId(), sender.getId(), FriendRequestStatus.PENDING);

        if (existsPending) {
            throw new BadRequestException("Friend request already exists");
        }

        // Check existing friendship
        if (friendRepository.existsByUserIdAndFriendId(sender.getId(), receiver.getId())) {
            throw new BadRequestException("Already friends");
        }

        FriendRequestEntity entity = FriendRequestEntity.builder()
                .sender(sender)
                .receiver(receiver)
                .status(FriendRequestStatus.PENDING)
                .build();

        FriendRequestEntity saved = friendRequestRepository.save(entity);

        return FriendRequestResponse.builder()
                .id(saved.getId())
                .senderId(sender.getId())
                .receiverId(receiver.getId())
                .status(saved.getStatus())
                .createdAt(saved.getCreatedAt())
                .build();
    }

    /**
     * Accept a friend request and create a DIRECT group.
     */
    @Transactional
    public Map<String, String> acceptFriendRequest(UserEntity currentUser, UUID requestId) {
        FriendRequestEntity req = friendRequestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("Friend request not found"));

        if (!req.getReceiver().getId().equals(currentUser.getId())) {
            throw new UnauthorizedException("Not authorized to accept this request");
        }

        if (req.getStatus() != FriendRequestStatus.PENDING) {
            throw new BadRequestException("Friend request already processed");
        }

        req.setStatus(FriendRequestStatus.ACCEPTED);
        friendRequestRepository.save(req);

        // Create bidirectional friendship
        friendRepository.save(FriendEntity.builder()
                .userId(req.getSender().getId()).friendId(req.getReceiver().getId()).build());
        friendRepository.save(FriendEntity.builder()
                .userId(req.getReceiver().getId()).friendId(req.getSender().getId()).build());

        // Create DIRECT group
        UserEntity sender = req.getSender();
        UserEntity receiver = req.getReceiver();

        GroupEntity directGroup = GroupEntity.builder()
                .name(sender.getName() + " & " + receiver.getName())
                .creator(currentUser)
                .type(GroupType.DIRECT)
                .isArchived(false)
                .build();
        GroupEntity savedGroup = groupRepository.save(directGroup);

        groupMemberRepository.save(GroupMemberEntity.builder()
                .groupId(savedGroup.getId()).userId(sender.getId()).build());
        groupMemberRepository.save(GroupMemberEntity.builder()
                .groupId(savedGroup.getId()).userId(receiver.getId()).build());

        return Map.of("message", "Friend request accepted", "direct_group_id", savedGroup.getId().toString());
    }

    /**
     * Reject a friend request.
     */
    @Transactional
    public FriendRequestResponse rejectFriendRequest(UserEntity currentUser, UUID requestId) {
        FriendRequestEntity req = friendRequestRepository.findById(requestId)
                .orElseThrow(() -> new ResourceNotFoundException("Friend request not found"));

        if (!req.getReceiver().getId().equals(currentUser.getId())) {
            throw new UnauthorizedException("Not authorized to reject this request");
        }

        if (req.getStatus() != FriendRequestStatus.PENDING) {
            throw new BadRequestException("Friend request already processed");
        }

        req.setStatus(FriendRequestStatus.REJECTED);
        friendRequestRepository.save(req);

        return FriendRequestResponse.builder()
                .id(req.getId())
                .senderId(req.getSender().getId())
                .receiverId(req.getReceiver().getId())
                .status(req.getStatus())
                .createdAt(req.getCreatedAt())
                .build();
    }

    /**
     * Get pending friend requests for the current user.
     */
    public List<FriendRequestResponse> getPendingRequests(UserEntity currentUser) {
        List<FriendRequestEntity> requests = friendRequestRepository
                .findByReceiverIdAndStatus(currentUser.getId(), FriendRequestStatus.PENDING);

        return requests.stream().map(req -> FriendRequestResponse.builder()
                .id(req.getId())
                .senderId(req.getSender().getId())
                .receiverId(req.getReceiver().getId())
                .status(req.getStatus())
                .createdAt(req.getCreatedAt())
                .senderName(req.getSender().getName())
                .senderEmail(req.getSender().getEmail())
                .build()).toList();
    }

    /**
     * Get list of friends for the current user.
     */
    public List<FriendResponse> getFriends(UserEntity currentUser) {
        List<FriendEntity> friends = friendRepository.findByUserId(currentUser.getId());

        return friends.stream().map(f -> FriendResponse.builder()
                .userId(f.getUserId())
                .friendId(f.getFriendId())
                .friendName(f.getFriend().getName())
                .friendEmail(f.getFriend().getEmail())
                .createdAt(f.getCreatedAt())
                .build()).toList();
    }

    /**
     * Get the DIRECT group for a specific friend.
     */
    public Map<String, Object> getDirectGroup(UserEntity currentUser, UUID friendId) {
        if (!friendRepository.existsByUserIdAndFriendId(currentUser.getId(), friendId)) {
            throw new ResourceNotFoundException("Friendship not found");
        }

        List<GroupEntity> allGroups = groupRepository.findAllGroupsByUserId(currentUser.getId());
        GroupEntity directGroup = null;

        for (GroupEntity group : allGroups) {
            if (group.getType() == GroupType.DIRECT &&
                    groupMemberRepository.existsByGroupIdAndUserId(group.getId(), friendId)) {
                directGroup = group;
                break;
            }
        }

        if (directGroup == null) {
            throw new ResourceNotFoundException("DIRECT group not found for this friendship");
        }

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("group_id", directGroup.getId());
        result.put("group_name", directGroup.getName());
        result.put("created_at", directGroup.getCreatedAt());
        return result;
    }

    /**
     * Remove a friend and delete the DIRECT group.
     */
    @Transactional
    public Map<String, String> removeFriend(UserEntity currentUser, UUID friendId) {
        boolean exists = friendRepository.existsByUserIdAndFriendId(currentUser.getId(), friendId) ||
                friendRepository.existsByUserIdAndFriendId(friendId, currentUser.getId());

        if (!exists) {
            throw new ResourceNotFoundException("Friendship not found");
        }

        // Find and delete DIRECT group
        List<GroupEntity> allGroups = groupRepository.findAllGroupsByUserId(currentUser.getId());
        for (GroupEntity group : allGroups) {
            if (group.getType() == GroupType.DIRECT &&
                    groupMemberRepository.existsByGroupIdAndUserId(group.getId(), friendId)) {
                // Delete all related data (cascade handles most, but be explicit for splits)
                List<ExpenseEntity> expenses = expenseRepository.findByGroupIdOrderByCreatedAtDesc(group.getId());
                for (ExpenseEntity expense : expenses) {
                    List<ExpenseItemEntity> items = expenseItemRepository.findByExpenseId(expense.getId());
                    for (ExpenseItemEntity item : items) {
                        List<SplitEntity> splits = splitRepository.findByExpenseItemId(item.getId());
                        splitRepository.deleteAll(splits);
                    }
                    expenseItemRepository.deleteAll(items);
                }
                expenseRepository.deleteAll(expenses);
                groupMemberRepository.deleteByGroupIdAndUserId(group.getId(), currentUser.getId());
                groupMemberRepository.deleteByGroupIdAndUserId(group.getId(), friendId);
                groupRepository.delete(group);
                break;
            }
        }

        // Delete bidirectional friendships
        friendRepository.deleteByUserIdAndFriendId(currentUser.getId(), friendId);
        friendRepository.deleteByUserIdAndFriendId(friendId, currentUser.getId());

        return Map.of("message", "Friend removed successfully");
    }
}
